File: C++/cloneGraph.cpp
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
    std::unordered_map <Node*, Node*> m_storage;
public:
    Node* cloneGraph(Node* node) {
        if(!node) return nullptr;
        if(m_storage[node]) return m_storage[node];
        m_storage[node] = new Node(node->val);
        for(int i = 0; i < node->neighbors.size(); ++i) {
             m_storage[node]->neighbors.push_back(cloneGraph(node->neighbors[i]));
        }
        return m_storage[node];       
    }
};

File: Python/clone-graph.py
# Time:  O(n)
# Space: O(n)



class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        dic = defaultdict(Node)
        dic[None] = None
        visited = set()
        dq = deque()
        if node:
            dq.append(node)
        while dq:
            cur = dq.popleft()
            if cur not in visited:
                visited.add(cur)
                dic[cur].val = cur.val
                for nd in cur.neighbors:
                    dic[cur].neighbors.append(dic[nd])
                    dq.append(nd)
        return dic[node]

